### PR TITLE
Chore: Query endpoint refactor

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
Get rid of using `legacydata.RequestHandler` in `HTTPServer`, `/api/tsdb/query` and `pkg/expr` with the goal of deprecating /api/tsdb/query and remove it completely eventually. This is the first step of cleaning up the HTTP API query endpoint.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
- Not sure all of these changes makes sense so would like some feedback to begin with.

### CODE REVIEW

File: pkg/api/metrics.go [hunk 1]
```diff
@@ -152,115 +122,170 @@ func toMacronResponse(qdr *backend.QueryDataResponse) response.Response {
...
-	}
+		if res.Error != nil {
+			dqr.Error = res.Error
+		}
 
-	resp, err := hs.legacyDataRequestHandler.HandleRequest(c.Req.Context(), ds, request)
-	if err != nil {
-		return response.Error(http.StatusInternalServerError, "Metric request error", err)
+		if res.Frames != nil {
+			dqr.Dataframes = legacydata.NewDecodedDataFrames(res.Frames)
```
Comment (marefr) [inline pkg/api/metrics.go:162]: Very much unsure if this is enough to make it backward compatible :shrug:

**Review by wbrowne — COMMENTED** [review]
I think the code looks great 🙌  Haven't taken for a spin or anything so can't speak to backwards compatibility yet

**Comment by ryantxu** [comment]
Looks good (just skimming code) -- is it reasonable to think we can target using expressions for everything?  The non-expressions case is just a query with only one datasource and no expressions

**Comment by ryantxu** [comment]
Also check: https://github.com/grafana/grafana/pull/41653  goal here is to to support the same request/response structure as the HTTP endpoint, but over websocket -- this will force query results followed by channel subscription to use the same host!

**Comment by marefr** [comment]
> is it reasonable to think we can target using expressions for everything?

I don't know the internal details of how SSE works so not sure it'll work out of the box, but should be possible to get it working. One potential difference with SSE is perhaps that there's always one query per data source (haven't confirmed this yet though) and regular queries is X queries per data source. 

Regarding #41653 extracts the query endpoint into a service makes sense. This PR though feels like a natural fit before his changes. @ryantxu @FZambia WDYT?

**Comment by FZambia** [comment]
> Regarding #41653 extracts the query endpoint into a service makes sense. This PR though feels like a natural fit before his changes. @ryantxu @FZambia WDYT?

Since another pr is mostly a POC this time – just proceed with this one and I'll update mine later. We need some time to evaluate whether RPC over WS is a reasonable thing to do.

**Comment by marefr** [comment]
I'm currently in the making of adding some tests

File: pkg/api/metrics.go [hunk 2]
```diff
@@ -16,251 +16,288 @@ import (
...
-			},
-			QueryType: q.QueryType,
-			JSON:      modelJSON,
-		})
+		return ds, nil
 	}
 
-	return req, nil
+	// Fallback to the datasourceId
+	id, err := query.Get("datasourceId").Int64()
```
Comment (wbrowne) [inline pkg/api/metrics.go:283]: This change as is currently suffers from the same problem that was recently fixed here https://github.com/grafana/grafana/pull/41446

**Review by wbrowne — CHANGES_REQUESTED** [review]
Found an issue (see comment)

File: pkg/api/metrics.go (hunk 2 above)
Comment (marefr) [reply pkg/api/metrics.go:283]: Nicely spotted. Weird we don't get a conflict for that one. Maybe handled conflicts badly earlier. I'll have a look at this first thing Monday

**Review by wbrowne — APPROVED** [review]
Tested and LGTM!

**Pull request merged**